### PR TITLE
feat: mini-css-extract-plugin

### LINKED ISSUE #3210: support mini-css-extract-plugin
### What problem does this feature solve?

there are lots of difference between experiments.css and mini-css-extract-plugin which makes migrate old application hard

### What does the proposed API of configuration look like?

compatible with mini-css-extract-plugin

### PR DESCRIPTION
**Pull request by JSerFeng**
## Summary

close #3210

## Require Documentation?

- [ ] No
- [ ] Yes, the corresponding rspack-website PR is \_\_

### CODE REVIEW

**Comment by JSerFeng** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/web-infra-dev/rspack/5072?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #4835** <a href="https://app.graphite.dev/github/pr/web-infra-dev/rspack/4835?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #4785** <a href="https://app.graphite.dev/github/pr/web-infra-dev/rspack/4785?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #4718** <a href="https://app.graphite.dev/github/pr/web-infra-dev/rspack/4718?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #5072** <a href="https://app.graphite.dev/github/pr/web-infra-dev/rspack/5072?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/web-infra-dev/rspack/5072?utm_source=stack-comment).

**Comment by JSerFeng** [comment]
!canary

**Comment by hardfist** [comment]
I personally think @rspack/plugin-mini-css-extract should be put into rspack.MiniCssExtractPlugin, since the rust part implementation and js implementation is deeply coupled and another js package will carry maintaince burden, cc @h-a-n-a @LingyuCoder @ahabhgk

**Comment by JSerFeng** [comment]
> I personally think @rspack/plugin-mini-css-extract should be put into rspack.MiniCssExtractPlugin

Agreed

**Comment by JSerFeng** [comment]
!canary

**Comment by JSerFeng** [comment]
!canary

**Comment by JSerFeng** [comment]
!canary

**Comment by JSerFeng** [comment]
!canary

**Comment by JSerFeng** [comment]
!canary

**Comment by JSerFeng** [comment]
!canary

**Comment by JSerFeng** [comment]
!canary

**Comment by JSerFeng** [comment]
!canary

**Comment by JSerFeng** [comment]
!canary

**Comment by JSerFeng** [comment]
!canary

**Comment by JSerFeng** [comment]
!canary

**Comment by JSerFeng** [comment]
!canary

**Comment by JSerFeng** [comment]
!canary

**Comment by JSerFeng** [comment]
!canary

File: packages/rspack/src/builtin-plugin/mini-css-extract/index.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	insert?: string | ((linkTag: HTMLLinkElement) => void);
+	attributes?: Record<string, string>;
+	linkType?: string | "text/css" | false;
+	runtime?: boolean;
+
+	// workaround for pathinto, deprecate this when rspack supports pathinfo
+	pathinfo?: boolean;
+}
+
+export class RspackCssExtractPlugin {
```
Comment (ahabhgk) [inline packages/rspack/src/builtin-plugin/mini-css-extract/index.ts:23]:
```suggestion
export class CssExtractRspackPlugin {
```

File: packages/rspack/src/exports.ts [hunk 2]
```diff
@@ -200,3 +200,5 @@ export type { SourceMapDevToolPluginOptions } from "./builtin-plugin";
 
 export { EvalDevToolModulePlugin } from "./builtin-plugin";
 export type { EvalDevToolModulePluginOptions } from "./builtin-plugin";
+
+export { RspackCssExtractPlugin } from "./builtin-plugin";
```
Comment (ahabhgk) [inline packages/rspack/src/exports.ts:204]:
```suggestion
export { CssExtractRspackPlugin } from "./builtin-plugin";
```

File: crates/rspack_binding_options/src/options/raw_builtins/mod.rs [hunk 3]
```diff
@@ -144,6 +146,7 @@ pub enum BuiltinPluginName {
   // rspack js adapter plugins
   // naming format follow XxxRspackPlugin
   JsLoaderRspackPlugin,
+  CssExtractPlugin,
```
Comment (ahabhgk) [inline crates/rspack_binding_options/src/options/raw_builtins/mod.rs:149]:
```suggestion
  CssExtractRspackPlugin,
```

**Review by ahabhgk — COMMENTED** [review]
Small naming issue

**Comment by JSerFeng** [comment]
!canary

**Review by ahabhgk — APPROVED** [review]

**Comment by hardfist** [comment]
@JSerFeng this need documentation update

File: packages/rspack/src/builtin-plugin/css-extract/index.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	pathinfo?: boolean;
+}
+
+export class CssExtractRspackPlugin {
+	static pluginName: string = "css-extract-rspack-plugin";
+	static loader: string = LOADER_PATH;
+
+	options: PluginOptions;
+
+	constructor(options?: PluginOptions) {
```
Comment (hardfist) [inline packages/rspack/src/builtin-plugin/css-extract/index.ts:29]: if this is not working with newTreeshaking:false, we should add warning when newTreeshaking is set to true @JSerFeng

**Pull request merged**